man/man1/notmuch-restore.1
.TH NOTMUCH-RESTORE 1 2013-12-30 "Notmuch 0.17"
.SH NAME
notmuch-restore \- restores the tags from the given file (see notmuch dump)

.SH SYNOPSIS

.B "notmuch restore"
.RB [ "--accumulate" ]
.RB [ "--format=(auto|batch-tag|sup)" ]
.RI "[ --input=<" filename "> ]"

.SH DESCRIPTION

Restores the tags from the given file (see
.BR "notmuch dump" ")."

The input is read from the given filename, if any, or from stdin.


Supported options for
.B restore
include
.RS 4
.TP 4
.B \-\-accumulate

The union of the existing and new tags is applied, instead of
replacing each message's tags as they are read in from the dump file.

.RE
.RS 4
.TP 4
.B \-\-format=(sup|batch-tag|auto)

Notmuch restore supports two plain text dump formats, with each line
specifying a message-id and a set of tags.
For details of the actual formats, see \fBnotmuch-dump\fR(1).

.RS 4
.TP 4
.B sup

The
.B sup
dump file format is specifically chosen to be
compatible with the format of files produced by sup-dump.
So if you've previously been using sup for mail, then the
.B "notmuch restore"
command provides you a way to import all of your tags (or labels as
sup calls them).

.RE
.RS 4
.TP 4
.B batch-tag

The
.B batch-tag
dump format is intended to more robust against malformed message-ids
and tags containing whitespace or non-\fBascii\fR(7) characters.  See
\fBnotmuch-dump\fR(1) for details on this format.

.B "notmuch restore"
updates the maildir flags according to tag changes if the
.B "maildir.synchronize_flags"
configuration option is enabled. See \fBnotmuch-config\fR(1) for
details.

.RE

.RS 4
.TP 4
.B auto

This option (the default) tries to guess the format from the
input. For correctly formed input in either supported format, this
heuristic, based the fact that batch-tag format contains no parentheses,
should be accurate.

.RE

.RE

.SH SEE ALSO

\fBnotmuch\fR(1), \fBnotmuch-config\fR(1), \fBnotmuch-count\fR(1),
\fBnotmuch-dump\fR(1), \fBnotmuch-hooks\fR(5),
\fBnotmuch-insert\fR(1), \fBnotmuch-new\fR(1),
\fBnotmuch-reply\fR(1), \fBnotmuch-search\fR(1),
\fBnotmuch-search-terms\fR(7), \fBnotmuch-show\fR(1),
\fBnotmuch-tag\fR(1)


NEWS
Notmuch 0.17 (2013-12-30)
=========================

Incompatible change in SHA1 computation
---------------------------------------

Previously on big endian architectures like sparc and powerpc the
computation of SHA1 hashes was incorrect. This meant that messages
with overlong or missing message-ids were given different computed
message-ids than on more common little endian architectures like i386
and amd64.  If you use notmuch on a big endian architecture, you are
strongly advised to make a backup of your tags using `notmuch dump`
before this upgrade. You can locate the affected files using something
like:

    notmuch dump | \
      awk '/^notmuch-sha1-[0-9a-f]{40} / \
        {system("notmuch search --exclude=false --output=files id:" $1)}'

Command-Line Interface
----------------------

New options to better support handling duplicate messages

  If more than one message file is associated with a message-id,
  `notmuch search --output=files` will print all of them. A new
  `--duplicate=N` option can be used to specify which duplicate to
  print for each message.

  `notmuch count` now supports `--output=files` option to output the
  number of files associated with matching messages. This may be
  bigger than the number of matching messages due to duplicates
  (i.e. multiple files having the same message-id).

Improved `notmuch new` performance for unchanged folders

  `notmuch new` now skips over unchanged folders more efficiently,
  which can substantially improve the performance of checking for new
  mail in some situations (like NFS-mounted Maildirs).

`notmuch reply --format=text` RFC 2047-encodes headers

  Previously, this used a mix of standard MIME encoding for the reply
  body and UTF-8 for the headers.  Now, the text format reply template
  RFC 2047-encodes the headers, making the output a valid RFC 2822
  message.  The JSON/sexp format is unchanged.

`notmuch compact` command

  The new `compact` command exposes Xapian's compaction
  functionality through a more convenient interface than
  `xapian-compact`. `notmuch compact` will compact the database to a
  temporary location, optionally backup the original database, and
  move the compacted database into place.

Emacs Interface
---------------

`notmuch-tree` (formerly `notmuch-pick`) has been added to mainline

  `notmuch-tree` is a threaded message view for the emacs
  interface. Each message is one line in the results and the thread
  structure is shown using UTF-8 box drawing characters (similar to
  Mutt's threaded view). It comes between search and show in terms of
  amount of output and can be useful for viewing both single threads
  and multiple threads.

  Using `notmuch-tree`

  The main key entries to notmuch tree are

  'z' enter a query to view using notmuch tree (works in hello,
      search, show and tree mode itself)

  'Z' view the current query in tree notmuch tree (works from search
      and show)

  Once in tree mode, keybindings are mostly in line with the rest of
  notmuch and are all viewable with '?' as usual.

  Customising `notmuch-tree`

  `notmuch-tree` has several customisation variables. The most
  significant is the first notmuch-tree-show-out which determines the
  behaviour when selecting a message (with RET) in tree view. By
  default tree view uses a split window showing the single message in
  the bottom pane. However, if this option is set then it views the
  whole thread in the complete window jumping to the selected message
  in the thread. In either case command-prefix selects the other option.

Tagging threads in search is now race-free

  Previously, adding or removing a tag from a thread in a search
  buffer would affect messages that had arrived after the search was
  performed, resulting in, for example, archiving messages that were
  never seen.  Tagging now affects only the messages that were in the
  thread when the search was performed.

`notmuch-hello` refreshes when switching to the buffer

  The hello buffer now refreshes whenever you switch to the buffer,
  regardless of how you get there.  You can disable automatic
  refreshing by customizing `notmuch-hello-auto-refresh`.

Specific mini-buffer prompts for tagging operations

  When entering tags to add or remove, the mini-buffer prompt now
  indicates what operation will be performed (e.g., "Tag thread", "Tag
  message", etc).

Built-in help improvements

  Documentation for many commands has been improved, as displayed by
  `notmuch-help` (usually bound to "?").  The bindings listed by
  `notmuch-help` also now include descriptions of prefixed commands.

Quote replies as they are displayed in show view

  We now render the parts for reply quoting the same way they are
  rendered for show. At this time, the notable change is that replies
  to text/calendar are now pretty instead of raw vcalendar.

Fixed inconsistent use of configured search order

  All ways of interactively invoking search now honor the value of
  `notmuch-search-oldest-first`.

Common keymap for notmuch-wide bindings

  Several key bindings have been moved from mode-specific keymaps to
  the single `notmuch-common-keymap`, which is inherited by each
  notmuch mode.  If you've customized your key bindings, you may want
  to move some of them to the common keymap.

The `notmuch-tag` function now requires a list of tag changes

  For users who have scripted the Emacs interface: the `notmuch-tag`
  API has changed.  Previously, it accepted either a list of tag
  changes or a space-separated string of tag changes.  The latter is
  no longer supported and the function now returns nothing.

Fixed `notmuch-reply` putting reply in primary selection

  On emacs 24 notmuch-reply used to put the cited text into the
  primary selection (which could lead to inadvertently pasting this
  cited text elsewhere). Now the primary-selection is not changed.

Fixed `notmuch-show` invisible part handling


notmuch-restore.c
/* notmuch - Not much of an email program, (just index and search)
 *
 * Copyright © 2009 Carl Worth
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see http://www.gnu.org/licenses/ .
 *
 * Author: Carl Worth <cworth@cworth.org>
 */

#include "notmuch-client.h"
#include "dump-restore-private.h"
#include "tag-util.h"
#include "string-util.h"

static regex_t regex;

/* Non-zero return indicates an error in retrieving the message,
 * or in applying the tags.  Missing messages are reported, but not
 * considered errors.
 */
static int
tag_message (unused (void *ctx),
	     notmuch_database_t *notmuch,
	     const char *message_id,
	     tag_op_list_t *tag_ops,
	     tag_op_flag_t flags)
{
    notmuch_status_t status;
    notmuch_message_t *message = NULL;
    int ret = 0;

    status = notmuch_database_find_message (notmuch, message_id, &message);
    if (status) {
	fprintf (stderr, "Error applying tags to message %s: %s\n",
		 message_id, notmuch_status_to_string (status));
	return 1;
    }
    if (message == NULL) {
	fprintf (stderr, "Warning: cannot apply tags to missing message: %s\n",
		 message_id);
	/* We consider this a non-fatal error. */
	return 0;
    }

    /* In order to detect missing messages, this check/optimization is
     * intentionally done *after* first finding the message. */
    if ((flags & TAG_FLAG_REMOVE_ALL) || tag_op_list_size (tag_ops))
	ret = tag_op_list_apply (message, tag_ops, flags);

    notmuch_message_destroy (message);

    return ret;
}

/* Sup dump output is one line per message. We match a sequence of
 * non-space characters for the message-id, then one or more
 * spaces, then a list of space-separated tags as a sequence of
 * characters within literal '(' and ')'. */

static int
parse_sup_line (void *ctx, char *line,
		char **query_str, tag_op_list_t *tag_ops)
{

    regmatch_t match[3];
    char *file_tags;
    int rerr;

    tag_op_list_reset (tag_ops);

    chomp_newline (line);

    /* Silently ignore blank lines */
    if (line[0] == '\0') {
	return 1;
    }

    rerr = xregexec (&regex, line, 3, match, 0);
    if (rerr == REG_NOMATCH) {
	fprintf (stderr, "Warning: Ignoring invalid sup format line: %s\n",
		 line);
	return 1;
    }

    *query_str = talloc_strndup_debug (ctx, line + match[1].rm_so,
				       match[1].rm_eo - match[1].rm_so);

    file_tags = talloc_strndup_debug (ctx, line + match[2].rm_so,
				      match[2].rm_eo - match[2].rm_so);

    char *tok = file_tags;
    size_t tok_len = 0;

    tag_op_list_reset (tag_ops);

    while ((tok = strtok_len (tok + tok_len, " ", &tok_len)) != NULL) {

	if (*(tok + tok_len) != '\0') {
	    *(tok + tok_len) = '\0';
	    tok_len++;
	}

	if (tag_op_list_append (tag_ops, tok, FALSE))
	    return -1;
    }

    return 0;

}

int
notmuch_restore_command (notmuch_config_t *config, int argc, char *argv[])
{
    notmuch_database_t *notmuch;
    notmuch_bool_t accumulate = FALSE;
    tag_op_flag_t flags = 0;
    tag_op_list_t *tag_ops;

    char *input_file_name = NULL;
    FILE *input = stdin;
    char *line = NULL;
    void *line_ctx = NULL;
    size_t line_size;
    ssize_t line_len;

    int ret = 0;
    int opt_index;
    int input_format = DUMP_FORMAT_AUTO;

    if (notmuch_database_open (notmuch_config_get_database_path (config),
			       NOTMUCH_DATABASE_MODE_READ_WRITE, &notmuch))
	return EXIT_FAILURE;

    if (notmuch_config_get_maildir_synchronize_flags (config))
	flags |= TAG_FLAG_MAILDIR_SYNC;

    notmuch_opt_desc_t options[] = {
	{ NOTMUCH_OPT_KEYWORD, &input_format, "format", 'f',
	  (notmuch_keyword_t []){ { "auto", DUMP_FORMAT_AUTO },
				  { "batch-tag", DUMP_FORMAT_BATCH_TAG },
				  { "sup", DUMP_FORMAT_SUP },
				  { 0, 0 } } },
	{ NOTMUCH_OPT_STRING, &input_file_name, "input", 'i', 0 },
	{ NOTMUCH_OPT_BOOLEAN,  &accumulate, "accumulate", 'a', 0 },
	{ 0, 0, 0, 0, 0 }
    };

    opt_index = parse_arguments (argc, argv, options, 1);
    if (opt_index < 0)
	return EXIT_FAILURE;

    if (! accumulate)
	flags |= TAG_FLAG_REMOVE_ALL;

    if (input_file_name) {
	input = fopen (input_file_name, "r");
	if (input == NULL) {
	    fprintf (stderr, "Error opening %s for reading: %s\n",
		     input_file_name, strerror (errno));
	    return EXIT_FAILURE;
	}
    }

    if (opt_index < argc) {
	fprintf (stderr, "Unused positional parameter: %s\n", argv[opt_index]);
	return EXIT_FAILURE;
    }

    tag_ops = tag_op_list_create (config);
    if (tag_ops == NULL) {
	fprintf (stderr, "Out of memory.\n");
	return EXIT_FAILURE;
    }

    do {
	line_len = getline (&line, &line_size, input);

	/* empty input file not considered an error */
	if (line_len < 0)
	    return EXIT_SUCCESS;

    } while ((line_len == 0) ||
	     (line[0] == '#') ||
	     /* the cast is safe because we checked about for line_len < 0 */
	     (strspn (line, " \t\n") == (unsigned)line_len));

    char *p;
    for (p = line; (input_format == DUMP_FORMAT_AUTO) && *p; p++) {
	if (*p == '(')
	    input_format = DUMP_FORMAT_SUP;
    }

    if (input_format == DUMP_FORMAT_AUTO)
	input_format = DUMP_FORMAT_BATCH_TAG;

    if (input_format == DUMP_FORMAT_SUP)
	if ( xregcomp (&regex,
		       "^([^ ]+) \\(([^)]*)\\)$",
		       REG_EXTENDED) )
	    INTERNAL_ERROR ("compile time constant regex failed.");

    do {
	char *query_string, *prefix, *term;

	if (line_ctx != NULL)
	    talloc_free (line_ctx);

	line_ctx = talloc_new (config);
	if (input_format == DUMP_FORMAT_SUP) {
	    ret = parse_sup_line (line_ctx, line, &query_string, tag_ops);
	} else {
	    ret = parse_tag_line (line_ctx, line, TAG_FLAG_BE_GENEROUS,
				  &query_string, tag_ops);

	    if (ret == 0) {
		ret = parse_boolean_term (line_ctx, query_string,
					  &prefix, &term);
		if (ret && errno == EINVAL) {
		    fprintf (stderr, "Warning: cannot parse query: %s (skipping)\n", query_string);
		    continue;
		} else if (ret) {
		    /* This is more fatal (e.g., out of memory) */
		    fprintf (stderr, "Error parsing query: %s\n",
			     strerror (errno));
		    ret = 1;
		    break;
		} else if (strcmp ("id", prefix) != 0) {
		    fprintf (stderr, "Warning: not an id query: %s (skipping)\n", query_string);
		    continue;
		}
		query_string = term;
	    }
	}

	if (ret > 0)
	    continue;

	if (ret < 0)
	    break;

	ret = tag_message (line_ctx, notmuch, query_string,
			   tag_ops, flags);
	if (ret)
	    break;

    }  while ((line_len = getline (&line, &line_size, input)) != -1);

    if (line_ctx != NULL)
	talloc_free (line_ctx);

    if (input_format == DUMP_FORMAT_SUP)
	regfree (&regex);

    if (line)
	free (line);

    notmuch_database_destroy (notmuch);

    if (input != stdin)
	fclose (input);

    return ret ? EXIT_FAILURE : EXIT_SUCCESS;
}
